### PR TITLE
Add custody by root request to sync

### PR DESCRIPTION
**Pull request by dapplion**
## Issue Addressed

Part of 
- https://github.com/sigp/lighthouse/issues/4983
- https://github.com/sigp/lighthouse/issues/6072

Build on / depends on
- https://github.com/sigp/lighthouse/pull/6274

## Proposed Changes

Adds a new higher order request to fetch all custody columns of a given block root.

It's a stateful request that may trigger one or more data_columns_by_root internally with features:
- per-column retry
- peer load balancing / prioritization of peer
- don't re-request peers that return failures for some time

### CODE REVIEW

**Comment by jimmygchen** [comment]
This was automatically closed by mergify as the upstream branch was merged. I'll reopen this PR against `unstable`.

**Comment by jimmygchen** [comment]
The refactoring on these PRs are making merging back to `das` really difficult, can we please avoid refactoring on these PRs, and instead do it against `das` branch? 

Some of the new function signatures require reworking parts of the existing code on `das` branch, so it would be easier if the refactor happens directly on `das` branch, otherwise we'd have to do the rework as part of the merge back to `das` which is really painful.

File: beacon_node/lighthouse_network/src/types/globals.rs
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/types/globals.rs:142]: Why not take the latest change from `das`branch and just use `good_custody_subnet_peer`?
This undos most the changes from #6218 - i understand selecting peers from entire peers is not the right solution, but even this doesn't fix it and just revert back to the previous version.

File: beacon_node/lighthouse_network/src/peer_manager/peerdb.rs
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/peer_manager/peerdb.rs:734]: Is all this change required? Why not just take the latest change from `das` branch? The change in #6218 fixes the custody lookup tests.

File: beacon_node/lighthouse_network/src/types/globals.rs
Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/types/globals.rs:142]: Also happy if you'd like to change how it works on the `das` branch but i think it's better to target rework on the `das` branch before we finish the merge to `unstable`.

File: beacon_node/network/src/sync/network_context/custody_by_root.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context/custody_by_root.rs:19]: For consistency:

```suggestion
type DataColumnSidecarList<E> = Vec<Arc<DataColumnSidecar<E>>>;
```

Also updated on the `das` branch `custody.rs`

File: beacon_node/network/src/sync/network_context/custody_by_root.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context/custody_by_root.rs:23]: What's the reason to change this from `CustodyId`? (on `das` branch)

File: beacon_node/network/src/sync/network_context/custody_by_root.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context/custody_by_root.rs:57]: Are we dumping the peer group?

`das` branch: 

```
type CustodyRequestResult<E> = Result<Option<(DataColumnSidecarList<E>, PeerGroup)>, Error>;
```

File: beacon_node/network/src/sync/network_context/custody_by_root.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/network_context/custody_by_root.rs:138]: This comment can be removed as we're no longer casting, although I'm not sure why we removed the `CustodyDataColumn` type

File: beacon_node/network/src/sync/block_lookups/tests.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/block_lookups/tests.rs:387]: Is this just for debugging? it doesn't actually assert anything but just prints

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs:969]: This parameter type isn't compatible with the `das` branch, and if we want to change this type i think we should change it there instead, as it would affect other part of the code (e.g. sampling) and make merging back and conga line difficult

File: beacon_node/network/src/sync/manager.rs
Comment (jimmygchen) [inline beacon_node/network/src/sync/manager.rs:989]: We already have this on `das` branch, is there a reason why you don't want it included in this PR? (this would diverge with `das` branch)

**Review by jimmygchen — CHANGES_REQUESTED** [review]
Nice, another 900+ lines down 🎉 
I've went through the code and I don't see anything impacting mainnet code path, so we should be good to go once comments are addressed and CI passes.

**Comment by jimmygchen** [comment]
@dapplion Is there any refactor in this PR that you'd like to merge to `unstable`? or shall we close this one?

**Comment by dapplion** [comment]
I ported a change to
- https://github.com/sigp/lighthouse/pull/6317

else nothing relevant so I'll close

**Pull request closed without merging**
